Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp

```diff
@@ -1442,13 +1444,30 @@ void NetworkProcessProxy::removeSession(WebsiteDataStore& websiteDataStore)
 {
     m_websiteDataStores.remove(websiteDataStore);
 
-    if (canSendMessage())
-        send(Messages::NetworkProcess::DestroySession { websiteDataStore.sessionID() }, 0);
+    if (canSendMessage()) {
+        if (auto completionHandler = m_removalCompletionHandlers.take(websiteDataStore.sessionID()))
+            sendWithAsyncReply(Messages::NetworkProcess::DestroySession { websiteDataStore.sessionID() }, [completionHandler = WTFMove(completionHandler)]() mutable {
+                return completionHandler({ });
+            });
+        else
+            sendWithAsyncReply(Messages::NetworkProcess::DestroySession { websiteDataStore.sessionID() }, [] { });
+    }
 
     if (m_websiteDataStores.isEmptyIgnoringNullReferences())
         defaultNetworkProcess() = nullptr;
 }
 
+void NetworkProcessProxy::setRemovalCompletionHandlerForWebsiteDataStore(WebsiteDataStore& websiteDataStore, CompletionHandler<void(String&&)>&& completionHandler)
+{
+    if (!m_websiteDataStores.contains(websiteDataStore))
+        return completionHandler("WebsiteDataStore is not used by network process"_s);
+
+    if (auto existingCompletionHandler = m_removalCompletionHandlers.take(websiteDataStore.sessionID()))
+        existingCompletionHandler("New completion handler is set"_s);
+
+    m_removalCompletionHandlers.set(websiteDataStore.sessionID(), WTFMove(completionHandler));
```

Comment:
Should use `add()` since you know the key is not present.

Reply:
Removed.